### PR TITLE
Add bindings for GtkTreeModelSort

### PR DESCRIPTION
**Pull request by ColonelThirtyTwo**
Like TreeModelFilter, this object wraps a TreeModel. This model can be sorted without
affecting the ordering of the original model, and can translate sorted TreePaths into
paths to the corresponding unsorted entries.

Example program:

```rust
extern crate gtk;
extern crate gio;
extern crate gdk;

use std::env::args;
use gio::ApplicationExt;
use gio::ApplicationExtManual;
use gtk::prelude::*;

fn build_ui(application: &gtk::Application) {
	let window = gtk::ApplicationWindow::new(application);
	window.set_title("Test Window");
	window.set_border_width(10);
	window.set_position(gtk::WindowPosition::Center);
	window.set_default_size(350, 70);
	window.connect_delete_event(move |win, _| {
		win.destroy();
		Inhibit(false)
	});
	
	let store = gtk::TreeStore::new(&[gtk::Type::String]);
	store.insert_with_values(None, None, &[0], &[&"One"]);
	store.insert_with_values(None, None, &[0], &[&"Two"]);
	store.insert_with_values(None, None, &[0], &[&"Three"]);
	store.insert_with_values(None, None, &[0], &[&"Four"]);
	
	let sortable_store = gtk::TreeModelSort::new(&store);
	
	let treeview = gtk::TreeView::new_with_model(&sortable_store);
	let column = gtk::TreeViewColumn::new();
	column.set_title("Value");
	column.set_clickable(true);
	column.set_sort_indicator(true);
	column.set_sort_column_id(0);
	
	let renderer = gtk::CellRendererText::new();
	column.pack_end(&renderer, true);
	column.add_attribute(&renderer, "text", 0);
	
	treeview.append_column(&column);
	
	treeview.connect_row_activated(move |_, path, _column| {
		let real_path = sortable_store.convert_path_to_child_path(&path)
			.expect("Sorted path does not correspond to real path");
		println!("Clicked on sorted: {:?}, real: {:?}", path.get_indices(), real_path.get_indices());
	});
	
	window.add(&treeview);
	window.show_all();
}

fn main() {
	let application = gtk::Application::new("com.github.basic",
		gio::ApplicationFlags::empty())
		.expect("Initialization failed...");
	
	application.connect_startup(|app| {
		build_ui(app);
	});
	application.connect_activate(|_| {});
	application.run(&args().collect::<Vec<_>>());
}
```

### CODE REVIEW

File: src/auto/tree_model_sort.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    fn convert_path_to_child_path(&self, sorted_path: &TreePath) -> Option<TreePath> {
+        unsafe {
+            from_glib_full(ffi::gtk_tree_model_sort_convert_path_to_child_path(self.to_glib_none().0, mut_override(sorted_path.to_glib_none().0)))
+        }
+    }
+
+    fn get_model(&self) -> Option<TreeModel> {
+        unsafe {
+            from_glib_none(ffi::gtk_tree_model_sort_get_model(self.to_glib_none().0))
```
Comment (EPashkin) [inline src/auto/tree_model_sort.rs]: This seems also need non-option return, as it created with non-option constructor.

File: src/auto/tree_model_sort.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        }
+    }
+
+    fn iter_is_valid(&self, iter: &mut TreeIter) -> bool {
+        unsafe {
+            from_glib(ffi::gtk_tree_model_sort_iter_is_valid(self.to_glib_none().0, iter.to_glib_none_mut().0))
```
Comment (EPashkin) [inline src/auto/tree_model_sort.rs]: `iter` IMHO don't need be mutable.

**Comment by EPashkin** [comment]
@ColonelThirtyTwo thanks, other that some nits looks good,

File: src/tree_model_sort.rs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+use glib::object::{Downcast, IsA};
+use glib::translate::*;
+
+use {TreeModel, TreeModelSort};
+
+impl TreeModelSort {
+    pub fn new<T: IsA<TreeModel>>(child_model: &T) -> TreeModelSort {
+        skip_assert_initialized!();
+        unsafe {
+            TreeModel::from_glib_none(ffi::gtk_tree_model_sort_new_with_model(child_model.to_glib_none().0))
```
Comment (GuillaumeGomez) [inline src/tree_model_sort.rs]: Wo, this code formatting looks super weird. Please change it to:

```rust
impl TreeModelSort {
    pub fn new<T: IsA<TreeModel>>(child_model: &T) -> TreeModelSort {
        skip_assert_initialized!();
        unsafe {
            TreeModel::from_glib_none(
                ffi::gtk_tree_model_sort_new_with_model(
                    child_model.to_glib_none().0)
                ).downcast_unchecked()
        }
    }
}
```

**Comment by GuillaumeGomez** [comment]
Two remaining nits and it's good to go, thanks! (oh and once merged, please add your code as example in the `gtk-rs/examples` repository!)

File: src/tree_model_sort.rs (hunk 3 above)
Comment (ColonelThirtyTwo) [reply src/tree_model_sort.rs]: I'll change this. FYI I mostly copied that from `TreeModelFilter`.

Comment (GuillaumeGomez) [reply src/tree_model_sort.rs]: If the code looks like this, I'm ashamed we accepted it. :O

**Comment by ColonelThirtyTwo** [comment]
@GuillaumeGomez I've addressed the issues. Can you take a look at this patch again?

**Comment by GuillaumeGomez** [comment]
There is still [this comment](https://github.com/gtk-rs/gtk/pull/721/files#r223193374) which is unresolved.

**Comment by sdroege** [comment]
@ColonelThirtyTwo Are you planning to update this PR?

**Comment by GuillaumeGomez** [comment]
Closing in favor of #777.

**Pull request closed without merging**